README.md
# RNCryptor

Cross-language AES Encryptor/Decryptor [data
format](https://github.com/RNCryptor/RNCryptor-Spec/blob/master/RNCryptor-Spec-v3.md).
 
The primary target is Objective-C, but implementations are available in
[C++](https://github.com/RNCryptor/RNCryptor-cpp),
[C#](https://github.com/RNCryptor/RNCryptor-cs),
[Java](https://github.com/RNCryptor/JNCryptor),
[PHP](https://github.com/RNCryptor/RNCryptor-php),
[Python](https://github.com/RNCryptor/RNCryptor-python),
[Javascript](https://github.com/RNCryptor/rncryptor-js),
and [Ruby](https://github.com/RNCryptor/ruby_rncryptor).

The data format includes all the metadata required to securely implement AES
encryption, as described in ["Properly encrypting with AES with
CommonCrypto,"](http://robnapier.net/aes-commoncrypto) and [*iOS 6
Programming Pushing the Limits*](http://iosptl.com), Chapter 15. Specifically,
it includes:

* AES-256 encryption
* CBC mode
* Password stretching with PBKDF2
* Password salting
* Random IV
* Encrypt-then-hash HMAC

## Basic Objective-C Usage

The most common in-memory use case is as follows:

``` objc
NSData *data = [@"Data" dataUsingEncoding:NSUTF8StringEncoding];
NSError *error;
NSData *encryptedData = [RNEncryptor encryptData:data
                                   	withSettings:kRNCryptorAES256Settings
                                          password:aPassword
                                             error:&error];
```

This generates an `NSData` including a header, encryption salt, HMAC salt, IV,
ciphertext, and HMAC. To decrypt this bundle:

``` objc
NSData *decryptedData = [RNDecryptor decryptData:encryptedData
                                    withPassword:aPassword
                                           error:&error];
```

Note that `RNDecryptor` does not require settings. These are read from the
header.

## Asynchronous use

`RNCryptor suports asynchronous use, specifically designed to work with
`NSURLConnection. This is also useful for cases where the encrypted or decrypted
`data will not comfortably fit in memory. If the data will comfortably fit in
`memory, asynchronous operation is best acheived using dispatch_async().

To operate in asynchronous mode, you create an `RNEncryptor` or `RNDecryptor`,
providing it a handler. This handler will be called as data is encrypted or
decrypted. As data becomes available, call `addData:`. When you reach the end of
the data call `finish`.

``` objc
- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
  [self.encryptedData addData:[self.cryptor addData:data]];
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
  [self.cryptor finish];
}

// Other connection delegates

- (void)decryptionDidFinish {
  if (self.cryptor.error) {
    // An error occurred. You cannot trust encryptedData at this point
  }
  else {
    // self.encryptedData is complete. Use it as you like
  }
  self.encryptedData = nil;
  self.cryptor = nil;
  self.connection = nil;
}

- (void)decryptRequest:(NSURLRequest *)request {
  self.encryptedData = [NSMutableData data];
  self.connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
  self.cryptor = [[RNDecryptor alloc] initWithPassword:self.password
                                               handler:^(RNCryptor *cryptor, NSData *data) {
                                                   [self.decryptedData appendData:data];
                                                   if (cryptor.isFinished) {
                                                     [self decryptionDidFinish];
                                                   }
                                                 }];
}
```

## Async and Streams

When performing async operations on streams, the data can come very quickly
(particularly if you're reading from a local file). If you use RNCryptor in a
naïve way, you'll queue a work blocks faster than the engine can process them
and your memory usage will spike. This is particularly true if there's only one
core, such as on an iPad 1. The solution is to only dispatch new work blocks as
the previous work blocks complete.

``` objc
// Make sure that this number is larger than the header + 1 block.
// 33+16 bytes = 49 bytes. So it shouldn't be a problem.
int blockSize = 32 * 1024;

NSInputStream *cryptedStream = [NSInputStream inputStreamWithFileAtPath:@"C++ Spec.pdf"];
NSOutputStream *decryptedStream = [NSOutputStream outputStreamToFileAtPath:@"/tmp/C++.crypt" append:NO];

[cryptedStream open];
[decryptedStream open];

// We don't need to keep making new NSData objects. We can just use one repeatedly.
__block NSMutableData *data = [NSMutableData dataWithLength:blockSize];
__block RNEncryptor *decryptor = nil;

dispatch_block_t readStreamBlock = ^{
  [data setLength:blockSize];
  NSInteger bytesRead = [cryptedStream read:[data mutableBytes] maxLength:blockSize];
  if (bytesRead < 0) {
    // Throw an error
  }
  else if (bytesRead == 0) {
    [decryptor finish];
  }
  else {
    [data setLength:bytesRead];
    [decryptor addData:data];
    NSLog(@"Sent %ld bytes to decryptor", (unsigned long)bytesRead);
  }
};

decryptor = [[RNEncryptor alloc] initWithSettings:kRNCryptorAES256Settings
                                         password:@"blah"
                                          handler:^(RNCryptor *cryptor, NSData *data) {
                                            NSLog(@"Decryptor recevied %ld bytes", (unsigned long)data.length);
                                            [decryptedStream write:data.bytes maxLength:data.length];
                                            if (cryptor.isFinished) {
                                              [decryptedStream close];
                                              // call my delegate that I'm finished with decrypting
                                            }
                                            else {
                                              // Might want to put this in a dispatch_async(), but I don't think you need it.
                                              readStreamBlock();
                                            }
                                          }];

// Read the first block to kick things off    
readStreamBlock();
```

I'll eventually get this into the API to simplify things. See [Cyrille's SO
question](http://stackoverflow.com/a/14586231/97337) for more discussion. Pull
requests welcome.

## Building

Comes packaged as a static library, but the source files can be dropped into any
project. The OpenSSL files are not required.

Requires `Security.framework`.

Supports 10.6+ and iOS 4+.

The current file format is v3. To read v1 files (see Issue #44), you need to set the compile-time macro `RNCRYPTOR_ALLOW_V1_BAD_HMAC`. It is not possible to write v1 files anymore.

## Design considerations

`RNCryptor` has several design goals, in order of importance:

### Easy to use correctly for most common use cases

The most critical concern is that it be easy for non-experts to use `RNCryptor` correctly. A framework that is more secure, but requires a steep learning curve on the developer will either be not used, or used incorrectly. Whenever possible, a single line of code should "do the right thing" for the most common cases.

This also requires that it fail correctly and provide good errors.

### Reliance on CommonCryptor functionality

`RNCryptor` has very little "security" code. It relies as much as possible on the OS-provided CommonCryptor. If a feature does not exist in CommonCryptor, then it generally will not be provided in `RNCryptor`.

### Best practice security

Wherever possible within the above constraints, the best available algorithms
are applied. This means AES-256, HMAC+SHA1, and PBKDF2:

* AES-256. While Bruce Schneier has made some interesting recommendations
regarding moving to AES-128 due to certain attacks on AES-256, my current
thinking is in line with [Colin
Percival](http://www.daemonology.net/blog/2009-07-31-thoughts-on-AES.html).
PBKDF2 output is effectively random, which should negate related-keys attacks
against the kinds of use cases we're interested in.

* AES-CBC mode. This was a somewhat complex decision, but the ubiquity of CBC
outweighs other considerations here. There are no major problems with CBC mode,
and nonce-based modes like CTR have other trade-offs. See ["Mode changes for
RNCryptor"](http://robnapier.net/blog/mode-rncryptor) for more details on this
decision.

* Encrypt-then-MAC. If there were a good authenticated AES mode on iOS (GCM for
instance), I would probably use that for its simplicity. Colin Percival makes
[good arguments for hand-coding an encrypt-than-
MAC](http://www.daemonology.net/blog/2009-06-24-encrypt-then-mac.html) rather
than using an authenticated AES mode, but in RNCryptor mananging the HMAC
actually adds quite a bit of complexity. I'd rather the complexity at a more
broadly peer-reviewed layer like CommonCryptor than at the RNCryptor layer. But
this isn't an option, so I fall back to my own Encrypt-than-MAC.

* HMAC+SHA256. No surprises here.

* PBKDF2. While bcrypt and scrypt may be more secure than PBKDF2, CommonCryptor
only supports PBKDF2. [NIST also continues to recommend
PBKDF2](http://security.stackexchange.com/questions/4781/do-any-security-experts-recommend-bcrypt-for-password-storage). We use 10k rounds of PBKDF2
which represents about 80ms on an iPhone 4.

### Code simplicity

`RNCryptor endeavors to be implemented as simply as possible, avoiding tricky
`code. It is designed to be easy to read and code review.

### Performance

Performance is a goal, but not the most important goal. The code must be secure
and easy to use. Within that, it is as fast and memory-efficient as possible.

### Portability

Without sacrificing other goals, it is preferable to read the output format of
`RNCryptor` on other platforms.

## Version History

* v2.2 Switches to file format v3 to deal with Issue #77.
* v2.1 Switches to file format v2 to deal with Issue #44.
* v2.0 adds asynchronous modes.
* v2.1 backports `RNCryptor` to older versions of Mac OS X (and possibly iOS).


## LICENSE

Except where otherwise indicated in the source code, this code is licensed under
the MIT License:

```
Permission is hereby granted, free of charge, to any person obtaining a 
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:
 
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
```

Portions of this code, indicated in the source, are licensed under the following
license:

```
/*-
 * Copyright (c) 2008 Damien Bergamini <damien.bergamini@free.fr>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
```

Portions of this code, indicated in the source, are licensed under the APSL
license:

```
/*
 * Copyright (c) 2006-2010 Apple Inc. All Rights Reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 *
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 *
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 *
 * @APPLE_LICENSE_HEADER_END@
 */
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [5.1.0](https://github.com/RNCryptor/RNCryptor/releases/tag/5.1.0)

* Update to Swift 5
* Uses CommonCrypto modular headers (available since 4.2)

## [5.0.3](https://github.com/RNCryptor/RNCryptor/releases/tag/5.0.3)

*  #272. Correct warnings related to incorrect memory access. The previous code was invalid Swift (it accessed a value inside its own `withUnsafeBytes` block), but happened to work.
* Removes a `Data.init` work-around that is no longer needed.

## [5.0.2](https://github.com/RNCryptor/RNCryptor/releases/tag/5.0.2)

* Update Swift Package Manager support
* Update Linux support
* Add tvOS support
* Update to Xcode 9.1 settings
* Update to Swift 4 (no code changes; completely backward compatible)
* Remove code coverage from Release builds (necessary for Carthage)

## [5.0.1](https://github.com/RNCryptor/RNCryptor/releases/tag/5.0.1)

* Perform final decryption after HMAC validation. Fixes #185. This improves security very slightly. There is no known or proposed attack that could exploit the previous behavior, but where possible it is better to validate prior to decryption.
* Fix CocoaPods

## [5.0.0](https://github.com/RNCryptor/RNCryptor/releases/tag/5.0.0) Swift 3.0. Drop ObjC bridge.

* Rework for Swift 3
* Drops ObjC bridging. There are too many problems with the Swift compiler at this point to support this cleanly. ObjC code should use [RNCryptor-ObjC](https://github.com/RNCryptor/RNCryptor-ObjC).
* #159 Add WatchOS support to Podspec
* Better debugging output in case of unexpected cryptor failure
* Added Examples/KeyDerivation to explain how to manually derive keys

## [4.0.0](https://github.com/RNCryptor/RNCryptor/releases/tag/4.0.0) - Complete rewrite in Swift with ObjC bridging

* No changes from beta except version bump

## [4.0.0-beta.1](https://github.com/RNCryptor/RNCryptor/releases/tag/4.0.0-beta.1) - Complete rewrite in Swift with ObjC bridging

## [3.0.1](https://github.com/RNCryptor/RNCryptor/releases/tag/RNCryptor-3.0.1) - CocoaPods fixes (Current ObjC Version)

### Fixed
* Removed private headers from CocoaPods

## [3.0.0](https://github.com/RNCryptor/RNCryptor/releases/tag/RNCryptor-3.0.0) - Remove OpenSSL

### Breaking
* Remove OpenSSL support. This has been moved to [rnapier/RNOpenSSLCryptor](https://github.com/rnapier/RNOpenSSLCryptor).

### Added
* Integrates with Swift as a framework
* Remove warnings on OS X 10.8

## [2.2](https://github.com/RNCryptor/RNCryptor/releases/tag/RNCryptor-2.2) - Data format v3

Version 2.2 is a fairly large release. It's been almost a year since 2.1 came out, and there are many small and large bug fixes.

V2.2 updates the file format from 2 to 3. It will read format 2 files, but will only write format 3. These are not readable by RNCryptor v2.1. See Issue #77 for details. The PHP, Python, and Ruby implementations also write format 3 and read format 2 or 3.

### Security

* Issue #91:  Use constant time HMAC comparisons to avoid timing attacks
* Issue #77: KeyForPassword() broken for multi-byte passwords (UTF-8)

### Changes

* Improved PHP, Python, and Ruby implementations
* Improved test cases, with test vectors
* Issue #76: Support OSX in podspec
* Resolved static analyzer warnings
* Ensure compatibility with iOS 4.2
* Accept settings to RNDecryptor (Issue #65)
* Copy password rather than retain it (Issue #64)
* Crash when reading v1 header


php/decrypt.php
<?php

//
//
//  AES128-apis.php
//
//  RNCryptor PHP BackEnd Script
//  Using kCCAlgorithmAES128  
//  Advanced Encryption Standard, 128-bit block
//
//  Copyright (c) 2013 Guysung Kim
//
//
//

require_once("crypt-common.inc");

define('IV_SIZE', 24);
define('SALT_SIZE', 12);
define('HMAC_SIZE', 44);
$gPassword = "myPassword";


/*
 *
 *  Based on https://github.com/rnapier/RNCryptor
 *  
 *  Argument: $b64_data - this is base64encoded data following the data format below.
 *
 *  Data Format see https://github.com/rnapier/RNCryptor/wiki/Data-Format for details.
 * 
 *  version (1 byte): Data format version. Currently 2.
 *  options (1 byte): bit 0 - uses password
 *  encryptionSalt (8 bytes): iff option includes "uses password"
 *  HMACSalt (8 bytes): iff options includes "uses password"
 *  IV (16 bytes)
 *  ciphertext (variable) -- Encrypted with CTR mode in v1.x, CBC mode in 2.0.
 *  HMAC (32 bytes)
 *
 *  Returns: Decrypted data.
 *
 *  // Corresponding iOS encrypt codes example.
 *  NSData *encryptedData = [RNEncryptor encryptData:srcData
 *                                      withSettings:kRNCryptorAES256Settings
 *                                          password:@"password"
 *                                             error:&error];
 *       
 *  NSString *encryptedDataBase64Final = [encryptedData base64EncodedString];
 *
 *  //If no error we send the post, voila!
 *  if (!error) {  
 *           NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
 *           // Post enctypted
 *           [params setObject:encryptedDataBase64Final forKey:@"JSONEncryptedBase64"];    
 *           // Do Post!
 *           [[RKClient sharedClient] post:postResourcePath params:params delegate:self];
 *   }
 *
 */
function decrypt_data($b64_data) {
    global $gPassword;

    // kRNCryptorAES256Settings 
    $algorithm = MCRYPT_RIJNDAEL_128;
    $key_size = 32;
    $mode = MCRYPT_MODE_CBC;
    $pbkdf2_iterations = 10000;
    $pbkdf2_prf = 'sha1';
    $hmac_algorithm = 'sha256';

    // back to binary              
    $bin_data = base64_decode($b64_data);
    // extract salt
    $salt = substr($bin_data, 2, 8);
    // extract HMAC salt
    $hmac_salt = substr($bin_data, 10, 8);
    // extract IV
    $iv = substr($bin_data, 18, 16);
    // extract data
    $data = substr($bin_data, 34, strlen($bin_data) - 34 - 32);
    $dataWithoutHMAC = chr(2).chr(1).$salt.$hmac_salt.$iv.$data;
    // extract HMAC
    $hmac = substr($bin_data, strlen($bin_data) - 32);
    // make HMAC key
    $hmac_key = hash_pbkdf2($pbkdf2_prf, $gPassword, $hmac_salt, $pbkdf2_iterations, $key_size, true);
    // make HMAC hash
    $hmac_hash = hash_hmac($hmac_algorithm, $dataWithoutHMAC , $hmac_key, true);
    // check if HMAC hash matches HMAC  
    if($hmac_hash != $hmac) {
        echo "HMAC mismatch".$nl.$nl.$nl;
        return false;
    }
    // make data key
    $key = hash_pbkdf2($pbkdf2_prf, $gPassword, $salt, $pbkdf2_iterations, $key_size, true);

    // decrypt
    $cypher = mcrypt_module_open(MCRYPT_RIJNDAEL_128, '', MCRYPT_MODE_CBC, '');
    // initialize encryption handle
    if (mcrypt_generic_init($cypher, $key, $iv) != -1) {
            // decrypt
            $decrypted = mdecrypt_generic($cypher, $data);

            // http://www.php.net/manual/en/function.mdecrypt-generic.php
            // We found that sometimes the resulting padding is not null characters "\0" but rather one of several control characters.
            // If you know your data is not supposed to have any trailing control characters "as we did" you can strip them like so.
            $decrypted = preg_replace( "/\p{Cc}*$/u", "", $decrypted );

            // clean up
            mcrypt_generic_deinit($cypher);
            mcrypt_module_close($cypher);

            return trim($decrypted);
    }
    return false;
}
                                                                                                                                                                                                                          

/*
 *	// Corresponding iOS encrypt codes example.
 *	// Use the CPCryptController from https://github.com/iosptl/ios6ptl/blob/master/ch15/CryptPic/CryptPic/RNCryptManager.m
 *	if (! [[CPCryptController sharedController] encryptData:srcData password:kRNCryptorPasswd error:&error] ) {
 *		NSLog(@"Could not encrypt data: %@", error);
 *	}
 *	
 *	// Do Base64 encoding for AES128 encrypted data
 *	NSString *saltBase64 = [[CPCryptController sharedController].salt base64EncodedString];
 *	NSString *HMACSaltBase64 = [[CPCryptController sharedController].HMACSalt base64EncodedString];
 *	NSString *ivBase64 = [[CPCryptController sharedController].iv base64EncodedString];
 *	NSString *encryptedDataBase64 = [[CPCryptController sharedController].encryptedData base64EncodedString];
 *	NSString *HMACBase64 = [[CPCryptController sharedController].HMAC base64EncodedString];
 *	
 *	// Build the final format
 *	NSString *encryptedDataBase64Final = [NSString stringWithFormat:@"%@%@%@%@%@",saltBase64, HMACSaltBase64, ivBase64, encryptedDataBase64, HMACBase64];
 *	DebugLog(@"encryptedDataBase64Final: %@",encryptedDataBase64Final );
 *	
 *	//If no error we send the post, voila!
 *	if (!error) {
 *		
 *		NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
 *		
 *		// Post enctypted
 *		[params setObject:encryptedDataBase64Final forKey:@"json_encrypted_base64"];
 *		
 *		// Do Post!
 *		[[RKClient sharedClient] post:postResourcePath params:params delegate:self];
 *		
 *	}
 */ 

function CPCryptController_decrypt_data($data) {
  global $gPassword;

  /* kRNCryptorAES256Settings */
  $algorithm = MCRYPT_RIJNDAEL_128;
  $key_size = 16;
  $mode = MCRYPT_MODE_CBC;
  $pbkdf2_iterations = 10000;
  $pbkdf2_prf = 'sha1';
  $hmac_algorithm = 'sha256';

  // extract salt
  $salt = substr($data, 0, SALT_SIZE); 
  $salt = base64_decode($salt);
  // Remove $salt from $data.
  $data = substr($data, SALT_SIZE); 

  // extract HMAC salt
  $hmac_salt = substr($data, 0, SALT_SIZE);
  $hmac_salt = base64_decode($hmac_salt);
  // Remove $hmac_salt from $data.
  $data = substr($data, SALT_SIZE); 

  // extract IV
  $iv = substr($data, 0, IV_SIZE);
  $iv = base64_decode($iv);
  // Remove $iv from $data.
  $data = substr($data, IV_SIZE); 

  // extract data
  $final_data = substr($data, 0, strlen($data) - HMAC_SIZE);
  $final_data = base64_decode($final_data);

  // extract HMAC
  $hmac = substr($data, strlen($data)-HMAC_SIZE);
  $hmac = base64_decode($hmac);

  // make HMAC key
  $hmac_key = hash_pbkdf2($pbkdf2_prf, $gPassword, $hmac_salt, $pbkdf2_iterations, $key_size, true);

  // make HMAC hash
  $hmac_hash = hash_hmac($hmac_algorithm, $final_data , $hmac_key, true);

  // check if HMAC hash matches HMAC : HMAC inteegrity checking
  if($hmac_hash != $hmac) {
   echo 'HMAC integrity check has been failed...';  
   return false;
  }

  // make data key
  $key = hash_pbkdf2($pbkdf2_prf, $gPassword, $salt, $pbkdf2_iterations, $key_size, true);

  $cypher = mcrypt_module_open(MCRYPT_RIJNDAEL_128, '', MCRYPT_MODE_CBC, '');
  // initialize encryption handle
  if (mcrypt_generic_init($cypher, $key, $iv) != -1) {
          // decrypt
          $decrypted = mdecrypt_generic($cypher, $final_data);

          // http://www.php.net/manual/en/function.mdecrypt-generic.php
          // We found that sometimes the resulting padding is not null characters "\0" but rather one of several control characters.
          // If you know your data is not supposed to have any trailing control characters "as we did" you can strip them like so.
          $decrypted = preg_replace( "/\p{Cc}*$/u", "", $decrypted ); 

          // clean up
          mcrypt_generic_deinit($cypher);
          mcrypt_module_close($cypher);

          return trim($decrypted);
  }

  return false;
}

?>